### PR TITLE
[main][feat] Adding "addCollateral" routine

### PR DESCRIPTION
**Pull request by spicysquid168**
## Description
Create `addCollateral` routine to enabling users to adding collateral even when the price dff between DEX and Oracle more than `maxPriceDiff` if `goRouge` is `true`. If users choose to omit Oracle check when adding collateral, the protocol will only check liquidity pool reserve consistency. It also allows users to add any size of collateral without the need of huge collateral to bring the position back down under `workFactor`.

## Why?
A lot of users requested this feature. This will helps to improve UX and usability of the protocol.

## ChangeLogs:
- ConfigurableInterestVaultConfig.sol
- SimpleVaultConfig.sol
- Vault.sol

### CODE REVIEW

**Review by sundayalpaca — APPROVED** [review]

File: contracts/6/protocol/Vault.sol [hunk 1]
```diff
@@ -257,6 +253,68 @@ contract Vault is IVault, ERC20UpgradeSafe, ReentrancyGuardUpgradeSafe, OwnableU
...
+      IWNativeRelayer(uint160(config.getWNativeRelayer())).withdraw(amount);
+      SafeToken.safeTransferETH(to, amount);
+    } else {
+      SafeToken.safeTransfer(token, to, amount);
+    }
+  }
+
+  /// @dev addCollateral to the given position.
+  /// @param id The ID of the position to add collaterals.
+  /// @param amount The anout of BTOKEN to be added to the position
```
Comment (almost-llamas) [inline contracts/6/protocol/Vault.sol:271]: Please add the definition for `goRogue`

File: contracts/6/protocol/Vault.sol [hunk 2]
```diff
@@ -257,6 +253,68 @@ contract Vault is IVault, ERC20UpgradeSafe, ReentrancyGuardUpgradeSafe, OwnableU
...
+    address worker = pos.worker;
+    uint256 healthBefore = IWorker(worker).health(id);
+    require(id < nextPositionID, "bad position id");
+    require(pos.owner == msg.sender, "!position owner");
+    require(healthBefore != 0, "!active position");
+    // 2. Book execution scope variables. Check if the given strategy is known add strat.
+    POSITION_ID = id;
+    (STRATEGY, ) = abi.decode(data, (address, bytes));
+    require(config.approvedAddStrategies(STRATEGY), "!approved strat");
+    // 3. If not goRouge than check worker stability, else only check reserve consistency.
```
Comment (almost-llamas) [inline contracts/6/protocol/Vault.sol:293]: than -> then

File: test/Vault_WaultSwap02.test.ts [hunk 3]
```diff
@@ -2753,63 +2648,415 @@ describe('Vault - WaultSwap02', () => {
...
+        // Position#1: Alice borrows 1 BTOKEN and supply another 1 BTOKEN
+        // After calling `work()`
+        // 2 BTOKEN needs to swap 0.0732967258967755614 BTOKEN to 0.042234424701074812 FTOKEN
+        // new reserve after swap will be 1.732967258967755614 BTOKEN 0.057765575298925188 FTOKEN
+        // based on optimal swap formula, BTOKEN-FTOKEN to be added into the LP will be 1.267032741032244386 BTOKEN + 0.042234424701074812 FTOKEN
+        // lp amount from adding liquidity will be (0.042234424701074812 / 0.057765575298925188) * 0.316227766016837933(first total supply) = 0.231205137369691323 LP
+        // new reserve after adding liquidity 2.999999999999999954 BTOKEN + 0.100000000000000000 FTOKEN
+        // ----------------
+        // BTOKEN-FTOKEN reserve = 2.999999999999999954 BTOKEN + 0.100000000000000000 FTOKEN
+        // BTOKEN-FTOKEN total supply = 0.547432903386529256 BTOKEN-FTOKEN LP
```
Comment (almost-llamas) [inline test/Vault_WaultSwap02.test.ts:2725]: Do we still need manual calculation note in the future, when we already use the swapHelper?

File: test/Vault_WaultSwap02.test.ts [hunk 4]
```diff
@@ -2753,63 +2648,415 @@ describe('Vault - WaultSwap02', () => {
...
+                KILL_FACTOR,
+                false,
+                false
+              );
+            });
+
+            it("should revert", async () => {
+              await revertReserveNotConsistent(true, addStrat.address);
+            });
+          });
```
Comment (almost-llamas) [inline test/Vault_WaultSwap02.test.ts:3057]: Should this `context("when reserve is inconsistent")` be in the same level as `context("when worker is unstable")`?
Now it is nested as its child context.

File: test/Vault_WaultSwap02.test.ts (hunk 3 above)
Comment (spicysquid168) [reply test/Vault_WaultSwap02.test.ts:2725]: No need manual cal if using swapHelper

File: test/Vault_WaultSwap02.test.ts (hunk 4 above)
Comment (almost-llamas) [reply test/Vault_WaultSwap02.test.ts:3057]: if this is the case, please refactor this on other Vaultish test file too

Comment (spicysquid168) [reply test/Vault_WaultSwap02.test.ts]: Moved to the same level as `context("when worker is unstable")`

File: test/helpers/swap.ts [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+
+  public async computeUnderlaying(
+    amount: BigNumber,
+    tokenA: string,
+    tokenB: string
+  ): Promise<[BigNumber, BigNumber, BigNumber, BigNumber]> {
+    const currLp = PancakePair__factory.connect(await this.factory.getPair(tokenA, tokenB), this.signer);
+    const reserveIdx = this.reserves.findIndex((r) => r.lp === currLp.address);
+    if (reserveIdx === -1) {
+      throw "not found reserve";
```
Comment (almost-llamas) [inline test/helpers/swap.ts:346]: Avoid throwing string, throw Error instance instead
```
throw new Error(...)

File: test/helpers/swap.ts [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+  public async computeAddLiquidity(
+    tokenA: string,
+    tokenB: string,
+    amountA: BigNumber,
+    amountB: BigNumber
+  ): Promise<[BigNumber, BigNumber, BigNumber]> {
+    const currLp = PancakePair__factory.connect(await this.factory.getPair(tokenA, tokenB), this.signer);
+    const reserveIdx = this.reserves.findIndex((r) => r.lp === currLp.address);
+    if (reserveIdx === -1) {
+      throw "not found reserve";
```
Comment (almost-llamas) [inline test/helpers/swap.ts:241]: Avoid throwing string, throw Error instance instead
```
throw new Error(...)

File: test/helpers/swap.ts [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+
+  public async computeRemoveLiquidiy(
+    tokenA: string,
+    tokenB: string,
+    liquidity: BigNumber
+  ): Promise<[amountA: BigNumber, amountB: BigNumber]> {
+    const currLp = PancakePair__factory.connect(await this.factory.getPair(tokenA, tokenB), this.signer);
+    const reserveIdx = this.reserves.findIndex((r) => r.lp === currLp.address);
+    if (reserveIdx === -1) {
+      throw "removeLiquidity: not found reserve";
```
Comment (almost-llamas) [inline test/helpers/swap.ts:214]: Avoid throwing string, throw Error instance instead
```
throw new Error(...)

File: test/helpers/swap.ts [hunk 8]
```diff
@@ -0,0 +1,361 @@
...
+    amtIn: BigNumber,
+    path: string[],
+    updateReserve: boolean
+  ): Promise<BigNumber[]> {
+    const amts = [amtIn];
+    for (let i = 1; i < path.length; i++) {
+      const currLp = PancakePair__factory.connect(await this.factory.getPair(path[i - 1], path[i]), this.signer);
+      const reserveIdx = this.reserves.findIndex((r) => r.lp === currLp.address);
+      if (reserveIdx === -1) {
+        throw "not found reserve";
```
Comment (almost-llamas) [inline test/helpers/swap.ts:118]: Avoid throwing string, throw Error instance instead
```
throw new Error(...)

File: test/helpers/swap.ts [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+  }
+
+  public async computeOneSidedOptimalSwap(amt: BigNumber, path: string[]): Promise<BigNumber> {
+    const currLp = PancakePair__factory.connect(
+      await this.factory.getPair(path[path.length - 1], path[0]),
+      this.signer
+    );
+    const reserveIdx = this.reserves.findIndex((r) => r.lp === currLp.address);
+    if (reserveIdx === -1) {
+      throw "not found reserve";
```
Comment (almost-llamas) [inline test/helpers/swap.ts:147]: Avoid throwing string, throw Error instance instead
```
throw new Error(...)

**Review by ricardo-alpaca — APPROVED** [review]

**Review by almost-llamas — APPROVED** [review]

**Pull request merged**